### ISSUE TITLE
BUG: Idle websocket connections go away

### ISSUE THREAD

**Issue by mmarcotte** [ustaxcourt/ef-cms#8109]
**Describe the Bug**

API Gateway times idle websocket connections out after 10 minutes of inactivity. This causes workflows that rely on websockets to fail.

This [stackoverflow reference offers a possible solution](https://stackoverflow.com/questions/58401092/api-gateway-websockets-dealing-with-10-minute-idle-connection-timeout) to establish a ping and pong connection to keep the connections alive.

This [proof of concept](https://github.com/ustaxcourt/ef-cms/compare/staging...websocket-reconnect-onclose) also helps re-establish a websocket connection if it becomes disconnected on the client's side. Some version of `.onclose(...)` handling should be included in the solution.

This is the primary cause for [the generate notices of trial session workflow failing](https://github.com/flexion/ef-cms/issues/9056).

**Business Impact/Reason for Severity**

**In which environment did you see this bug?**

Any

**Who were you logged in as?**

Docketclerk

**What were you doing when you discovered this bug? (Using the application, demoing, smoke tests, testing other functionality, etc.)**

Sending notices on large trial sessions

**To Reproduce**

1. Login to any environment
2. Do anything that doesn't involve sending websockets for at least 10 minutes.
3. Do something that involves websockets (sending notices for trial sessions)
4. Get the spinning wheel 

**Expected Behavior**

The websocket connection should remain active throughout the session.

**Actual Behavior**

The websocket connection goes away after 10 minutes of inactivity, and it is not reestablished. Notices to not get retrieved by the client from the server.

**Screenshots**

**Cause of Bug, If Known**

https://docs.aws.amazon.com/apigateway/latest/developerguide/limits.html
> Idle Connection Timeout | 10 minutes

**Process for Logging a Bug:**
* Complete the above information
* Add a severity tag (Critical, High Severity, Medium Severity or Low Severity). See below for priority definition. 

**Severity Definition:**
* Critical Defect
Blocks entire system's or module’s functionality
No workarounds available
Testing cannot proceed further without bug being fixed.

* High-severity Defect
Affects key functionality of an application
There's a workaround, but not obvious or easy
App behaves in a way that is strongly different from the one stated in the requirements

* Medium-severity Defect
A minor function does not behave in a way stated in the requirements.
Workaround is available and easy

* Low-severity Defect
Mostly related to an application’s UI
Doesn't need a workaround, because it doesn't impact functionality

## Definition of Ready for Bugs(Created 10-4-21)
Definition used: A failure or flaw in the system which produces an incorrect or undesired result that deviates from the expected result or behavior. (Note: Expected results are use cases that have been documented in past user stories as acceptance criteria and test cases, and do not include strange behavior unrelated to use cases.)

The following criteria must be met in order for the development team to begin work on the bug.

The bug must:
 - Be focused on solving a user problem
 - Contain data for all fields in the bug template, so the team can pick it up and begin working immediately

Process: If the unexpected results are new use cases that have been identified, but not yet built, new acceptance criteria and test cases should be captured in a new user story and prioritized by the product owner. 

If the Court is not able to reproduce the bug, add the “Unable to reproduce” tag. This will provide visibility into the type of support that may be needed by the Court. In the event that the Court cannot reproduce the bug, the Court will work with Flexion to communicate what type of troubleshooting help may be needed.

## Definition of Done (Updated 4-14-21)
**Product Owner**
 - [ ]  Bug fix has been validated in the Court's test environment

**Engineering**
 - [x] Automated test scripts have been written
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Deployed to the Court's test environment

**Comment by ttlenard** [comment]
Tested this morning and ran into the spinner issue again.  Here are some screen grabs: 
![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/191d5a0f-09a3-4abc-8b46-0b034101c336)

![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/02049c5e-ecae-4b2b-a199-d86c0161051c)

**Comment by ttlenard** [comment]
**Testing Results from 10-26-21**

**Did not work**
2/7/22 - Chicago,  Illinois - 75f2fbe1-1946-4697-b07a-12972d4f8f6c
1/31/22 - Los Angeles, California - 7e548830-7f66-4b80-95cf-a38f20464551

**Worked**
1/10/22 - Chicago, Illinois - e540f014-08cb-45b4-b04f-61c5a8ab9707
2/14/22 - San Francisco, California - 1af868ff-2d33-4216-b0ab-b453bf65f093

**Testing Results from 10-27-21**

**Worked**
1/10/22 - Los Angeles, California - 5b4ff198-623a-48dd-8b67-6dcefa3839a0
1/24/22 - Miami, Florida - a6e7de5e-cd78-4af9-9aa9-c522ab5f1570

**Comment by cholly75** [comment]
Closed ustaxcourt/ef-cms#8080 - duplicate

**Comment by ttlenard** [comment]
Here is a screen grab of the test results from 10-26, and which ones worked vs failed. 
![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/0bcff352-6789-4c80-af72-8795d169818e)

Looked at the logs with Cody and saw where we were both logged in as Petitions Clerk 1 role, and we may have had some overlapping requests?  Theory matches with the top one, but we didn't see overlap in the bottom one:
![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/51e2496e-8236-4183-9568-db76d30974ad)

Today (10-27-21), I did a total 7 tests with setting the calendar for large trial sessions, and 6 of them passed.  The one that failed was when Cody and I both initiated setting a calendar with the same login at similar times, so I think that is why that one failed.  

I also tested downloading cases from large trial sessions, and I never received a time out.  We are feeling good moving this bug to done for now.
